svgnative/ports/cairo/CairoSVGRenderer.cpp

```diff
@@ -0,0 +1,473 @@
+        return;
+    }
+    free((void *)blob_data);
+    throw ("image is broken, or not PNG or JPEG\n");
+}
+
+CairoSVGImageData::~CairoSVGImageData()
+{
+    cairo_surface_destroy( mImageData );
+}
+
+float CairoSVGImageData::Width() const
+{
+    if (!mImageData)
+        return 0;
+    return static_cast<float>( cairo_image_surface_get_width (mImageData) );
+}
+
+float CairoSVGImageData::Height() const
+{
+    if (!mImageData)
+        return 0;
+    return static_cast<float>( cairo_image_surface_get_height (mImageData) );
+}
+
+CairoSVGRenderer::CairoSVGRenderer()
+{
+}
+
+CairoSVGRenderer::~CairoSVGRenderer()
+{
+}
+
+inline cairo_path_t* getPathObjFromCairoSvgPath( const Path* path )
+{
+    cairo_t* cr = static_cast<const CairoSVGPath*>(path)->mPath;
+    return cairo_copy_path( cr );
+}
+
+inline int countPointsFromCairoSvgPath( const Path* path )
+{
+    cairo_path_t* cairoPath = getPathObjFromCairoSvgPath( path );
+    int r = countPointsFromCairoPath( cairoPath );
+    cairo_path_destroy( cairoPath );
+    return r;
+}
+
+inline cairo_path_t* getTransformedClippingPath( const ClippingPath* clippingPath )
+{
+    cairo_path_t* path = getPathObjFromCairoSvgPath( clippingPath->path.get() );
+    if (clippingPath->transform)
+    {
+        cairo_matrix_t matrix = static_cast<const CairoSVGTransform*>(clippingPath->transform.get())->mMatrix;
+        cairo_path_t* pathTransformed = (cairo_path_t*)malloc( sizeof( cairo_path_t ) );
+        pathTransformed->num_data = path->num_data;
+        pathTransformed->status = path->status;
+        pathTransformed->data = (cairo_path_data_t*)malloc( path->num_data * sizeof( cairo_path_data_t ) );
+        for (int i = 0; i < path->num_data; i += path->data[i].header.length )
+        {
+            pathTransformed->data[i].header = path->data[i].header;
+            for (int j = 1; j < path->data[i].header.length; j ++ )
+            {
+                double x = path->data[i+j].point.x;
+                double y = path->data[i+j].point.y;
+                cairo_matrix_transform_point( &matrix, &x, &y );
+                pathTransformed->data[i+j].point.x = x;
+                pathTransformed->data[i+j].point.y = y;
+            }
+        }
+        cairo_path_destroy( path );
+        return pathTransformed;
+    } else
+        return path;
+}
+
+void CairoSVGRenderer::Save(const GraphicStyle& graphicStyle)
+{
+    SVG_ASSERT( mCairo );
+    cairo_save ( mCairo );
+
+    if (graphicStyle.transform)
+        cairo_transform( mCairo, &(static_cast<CairoSVGTransform*>(graphicStyle.transform.get())->mMatrix));
+
+    if (graphicStyle.clippingPath && graphicStyle.clippingPath->path)
+    {
+        cairo_path_t* path = getTransformedClippingPath( graphicStyle.clippingPath.get() );
+        cairo_append_path( mCairo, path );
+        cairo_clip( mCairo );
+        cairo_path_destroy( path );
+        // FIXME: Cairo has no API to control winding/even-odd clipping. See
```

Comment:
according to https://www.cairographics.org/manual/cairo-cairo-t.html#cairo-set-fill-rule , clipping path itself cannot hold the rule for winding/even-odd. It would be reflected when filling is executed, and the rule can be set by cairo_set_fill_rule(). But this stage (Save() method), we cannot execute the filling.
How to resolve this problem? Should we push the current winding/even-odd rule to some stack?

Reply:
Ok, so why don't you set it right before calling cairo_clip? If needed you could hold the winding rule in a temp variable but it seems easier to just set it right before any clip/draw command.